File: server/src/nebula_api/routes/audit.py
"""Audit API routes."""

# Third-Party
from fastapi import APIRouter, Depends, Query, Request

# Local
from nebula_api.auth import require_auth
from nebula_api.response import paginated, success
from nebula_mcp.helpers import (
    list_audit_actors,
    list_audit_scopes,
    query_audit_log,
)

router = APIRouter()


@router.get("/")
async def list_audit_log(
    request: Request,
    auth: dict = Depends(require_auth),
    table: str | None = None,
    action: str | None = None,
    actor_type: str | None = None,
    actor_id: str | None = None,
    record_id: str | None = None,
    scope_id: str | None = None,
    limit: int = Query(50, le=200),
    offset: int = 0,
) -> dict:
    """List audit log entries with optional filters.

    Args:
        request: FastAPI request.
        auth: Auth context.
        table: Table name filter.
        action: Action filter.
        actor_type: Actor type filter.
        actor_id: Actor id filter.
        record_id: Record id filter.
        scope_id: Privacy scope filter.
        limit: Max rows.
        offset: Offset for pagination.

    Returns:
        Paginated audit log response.
    """
    pool = request.app.state.pool
    rows = await query_audit_log(
        pool,
        table,
        action,
        actor_type,
        actor_id,
        record_id,
        scope_id,
        limit,
        offset,
    )
    return paginated(rows, len(rows), limit, offset)


@router.get("/scopes")
async def list_scopes(
    request: Request,
    auth: dict = Depends(require_auth),
) -> dict:
    """List audit scopes with usage counts.

    Args:
        request: FastAPI request.
        auth: Auth context.

    Returns:
        List of scopes with counts.
    """
    pool = request.app.state.pool
    rows = await list_audit_scopes(pool)
    return success(rows)


@router.get("/actors")
async def list_actors(
    request: Request,
    auth: dict = Depends(require_auth),
    actor_type: str | None = None,
) -> dict:
    """List audit actors with activity counts.

    Args:
        request: FastAPI request.
        auth: Auth context.
        actor_type: Actor type filter.

    Returns:
        List of audit actors.
    """
    pool = request.app.state.pool
    rows = await list_audit_actors(pool, actor_type)
    return success(rows)

File: cli/src/internal/api/audit.go
package api

import "fmt"

// QueryAuditLog retrieves audit log entries with optional filters.
func (c *Client) QueryAuditLog(params QueryParams) ([]AuditEntry, error) {
	data, err := c.get(buildQuery("/api/audit", params))
	if err != nil {
		return nil, err
	}
	return decodeList[AuditEntry](data)
}

// QueryAuditLogWithPagination builds common audit params.
func (c *Client) QueryAuditLogWithPagination(
	tableName string,
	action string,
	actorType string,
	actorID string,
	recordID string,
	scopeID string,
	limit int,
	offset int,
) ([]AuditEntry, error) {
	params := QueryParams{}
	if tableName != "" {
		params["table"] = tableName
	}
	if action != "" {
		params["action"] = action
	}
	if actorType != "" {
		params["actor_type"] = actorType
	}
	if actorID != "" {
		params["actor_id"] = actorID
	}
	if recordID != "" {
		params["record_id"] = recordID
	}
	if scopeID != "" {
		params["scope_id"] = scopeID
	}
	params["limit"] = fmt.Sprintf("%d", limit)
	params["offset"] = fmt.Sprintf("%d", offset)
	return c.QueryAuditLog(params)
}

// ListAuditScopes retrieves privacy scopes with usage stats.
func (c *Client) ListAuditScopes() ([]AuditScope, error) {
	data, err := c.get("/api/audit/scopes")
	if err != nil {
		return nil, err
	}
	return decodeList[AuditScope](data)
}

// ListAuditActors retrieves audit actor summaries.
func (c *Client) ListAuditActors(actorType string) ([]AuditActor, error) {
	params := QueryParams{}
	if actorType != "" {
		params["actor_type"] = actorType
	}
	data, err := c.get(buildQuery("/api/audit/actors", params))
	if err != nil {
		return nil, err
	}
	return decodeList[AuditActor](data)
}
